From a9c2d915f1bab195dc6202bf62cbfa6e13b8ffda Mon Sep 17 00:00:00 2001
From: Daniel DeGrasse <ddegrasse@tenstorrent.com>
Date: Wed, 17 Sep 2025 16:35:34 -0500
Subject: [PATCH] scripts: twister: add failure_script parameter to run after
 test failure

Add "failure_script", a commandline or hardware map parameter that
allows users to execute a custom script after a testcase fails on a DUT.
This can be useful to dump device state after a test failure, especially
one that did not produce any console output.

Signed-off-by: Daniel DeGrasse <ddegrasse@tenstorrent.com>
---
 scripts/pylib/twister/twisterlib/environment.py |  5 +++++
 scripts/pylib/twister/twisterlib/handlers.py    |  6 ++++++
 scripts/pylib/twister/twisterlib/hardwaremap.py | 12 ++++++++++--
 scripts/schemas/twister/hwmap-schema.yaml       |  6 ++++++
 scripts/tests/twister/test_handlers.py          |  2 ++
 scripts/tests/twister/test_hardwaremap.py       |  4 ++++
 6 files changed, 33 insertions(+), 2 deletions(-)

diff --git a/scripts/pylib/twister/twisterlib/environment.py b/scripts/pylib/twister/twisterlib/environment.py
index f956f853c2b..41a3fc551e5 100644
--- a/scripts/pylib/twister/twisterlib/environment.py
+++ b/scripts/pylib/twister/twisterlib/environment.py
@@ -683,6 +683,11 @@ structure in the main Zephyr tree: boards/<vendor>/<board_name>/""")
                         before device handler open serial port and invoke runner.
                         """)

+    parser.add_argument("--failure-script",
+                        help="""specify a failures script. This will be executed
+                        on test failure.
+                        """)
+
     parser.add_argument(
         "--quarantine-list",
         action="append",
diff --git a/scripts/pylib/twister/twisterlib/handlers.py b/scripts/pylib/twister/twisterlib/handlers.py
index dc1ff258935..ce43c530061 100755
--- a/scripts/pylib/twister/twisterlib/handlers.py
+++ b/scripts/pylib/twister/twisterlib/handlers.py
@@ -522,6 +522,12 @@ class DeviceHandler(Handler):
     def make_dut_available(self, dut):
         if self.instance.status in [TwisterStatus.ERROR, TwisterStatus.FAIL]:
             dut.failures_increment()
+            if dut.failure_script:
+                logger.debug(f"Running failure script: {dut.failure_script}")
+                timeout = 30
+                if dut.script_param:
+                    timeout = dut.script_param.get("failure_script_timeout", timeout)
+                self.run_custom_script(re.split('[, ]', dut.failure_script), timeout=timeout)
         logger.debug(f"Release DUT:{dut.platform}, Id:{dut.id}, "
                      f"counter:{dut.counter}, failures:{dut.failures}")
         # get all DUTs with the same id
diff --git a/scripts/pylib/twister/twisterlib/hardwaremap.py b/scripts/pylib/twister/twisterlib/hardwaremap.py
index 74c384ab5d6..3b1d5381362 100644
--- a/scripts/pylib/twister/twisterlib/hardwaremap.py
+++ b/scripts/pylib/twister/twisterlib/hardwaremap.py
@@ -49,7 +49,8 @@ class DUT:
                  runner=None,
                  flash_timeout=60,
                  flash_with_test=False,
-                 flash_before=False):
+                 flash_before=False,
+                 failure_script=None):

         self.serial = serial
         self.baud = serial_baud or 115200
@@ -70,6 +71,7 @@ class DUT:
         self.post_script = post_script
         self.pre_script = pre_script
         self.script_param = script_param
+        self.failure_script = failure_script
         self.probe_id = None
         self.notes = None
         self.lock = Lock()
@@ -198,6 +200,7 @@ class HardwareMap:
                 self.add_device(self.options.device_serial,
                                 self.options.platform[0],
                                 self.options.pre_script,
+                                self.options.failure_script,
                                 False,
                                 baud=self.options.device_serial_baud,
                                 flash_timeout=self.options.device_flash_timeout,
@@ -209,6 +212,7 @@ class HardwareMap:
                 self.add_device(self.options.device_serial_pty,
                                 self.options.platform[0],
                                 self.options.pre_script,
+                                self.options.failure_script,
                                 True,
                                 flash_timeout=self.options.device_flash_timeout,
                                 flash_with_test=self.options.device_flash_with_test,
@@ -238,6 +242,7 @@ class HardwareMap:
         serial,
         platform,
         pre_script,
+        failure_script,
         is_pty,
         baud=None,
         flash_timeout=60,
@@ -248,6 +253,7 @@ class HardwareMap:
             platform=platform,
             connected=True,
             pre_script=pre_script,
+            failure_script=failure_script,
             serial_baud=baud,
             flash_timeout=flash_timeout,
             flash_with_test=flash_with_test,
@@ -291,6 +297,7 @@ class HardwareMap:
             product = dut.get('product')
             fixtures = dut.get('fixtures', [])
             connected = dut.get('connected') and ((serial or serial_pty) is not None)
+            failure_script = dut.get('failure_script')
             if not connected:
                 continue
             for plat in platforms:
@@ -309,7 +316,8 @@ class HardwareMap:
                               post_flash_script=post_flash_script,
                               script_param=script_param,
                               flash_timeout=flash_timeout,
-                              flash_with_test=flash_with_test)
+                              flash_with_test=flash_with_test,
+                              failure_script=failure_script)
                 new_dut.fixtures = fixtures
                 new_dut.counter = 0
                 self.duts.append(new_dut)
diff --git a/scripts/schemas/twister/hwmap-schema.yaml b/scripts/schemas/twister/hwmap-schema.yaml
index 142d4a1969b..f369a84f3f3 100644
--- a/scripts/schemas/twister/hwmap-schema.yaml
+++ b/scripts/schemas/twister/hwmap-schema.yaml
@@ -50,6 +50,9 @@ sequence:
       "pre_script":
         type: str
         required: false
+      "failure_script":
+        type: str
+        required: false
       "fixtures":
         type: seq
         required: false
@@ -77,3 +80,6 @@ sequence:
           "post_flash_timeout":
             type: int
             required: false
+          "failure_script_timeout":
+            type: int
+            required: false
diff --git a/scripts/tests/twister/test_handlers.py b/scripts/tests/twister/test_handlers.py
index e4710060a35..41938878ffd 100644
--- a/scripts/tests/twister/test_handlers.py
+++ b/scripts/tests/twister/test_handlers.py
@@ -1439,6 +1439,7 @@ def test_devicehandler_handle(
         pre_script='dummy pre script',
         post_script='dummy post script',
         post_flash_script='dummy post flash script',
+        failure_script='dummy failure script',
         flash_timeout=60,
         flash_with_test=True
     )
@@ -1491,6 +1492,7 @@ def test_devicehandler_handle(
         mock.call('dummy pre script', mock.ANY),
         mock.call('dummy post flash script', mock.ANY),
-        mock.call('dummy post script', mock.ANY)
+        mock.call('dummy post script', mock.ANY),
+        mock.call('dummy failure script', mock.ANY)
     ])

     if expected_reason:
diff --git a/scripts/tests/twister/test_hardwaremap.py b/scripts/tests/twister/test_hardwaremap.py
index 5dc61dc1d9e..81586d47c65 100644
--- a/scripts/tests/twister/test_hardwaremap.py
+++ b/scripts/tests/twister/test_hardwaremap.py
@@ -57,6 +57,7 @@ TESTDATA_1 = [
             'pre_script': 'dummy pre script',
             'post_script': 'dummy post script',
             'post_flash_script': 'dummy post flash script',
+            'failure_script': 'dummy failure script',
             'runner': 'dummy runner',
             'flash_timeout': 30,
             'flash_with_test': True,
@@ -64,6 +65,7 @@ TESTDATA_1 = [
                 'pre_script_timeout' : 30,
                 'post_flash_timeout' : 30,
                 'post_script_timeout' : 30,
+                'failure_script_timeout' : 30,
                 }
         },
         {
@@ -79,6 +81,7 @@ TESTDATA_1 = [
             'pre_script': 'dummy pre script',
             'post_script': 'dummy post script',
             'post_flash_script': 'dummy post flash script',
+            'failure_script': 'dummy failure script',
             'runner': 'dummy runner',
             'flash_timeout': 30,
             'flash_with_test': True,
@@ -86,6 +89,7 @@ TESTDATA_1 = [
                 'pre_script_timeout' : 30,
                 'post_flash_timeout' : 30,
                 'post_script_timeout' : 30,
+                'failure_script_timeout' : 30,
                 }
         },
         '<dummy platform (dummy product) on dummy serial>'
--
2.34.1
